File: Object-C/BlackFireIM/Example/ChatDetail/ChatRoom/BFEditTodInfoController.m
//
//  BFEditTodInfoController.m
//  BlackFireIM
//
//  Created by benny wang on 2021/11/8.
//

#import "BFEditTodInfoController.h"
#import "MSIMSDK-UIKit.h"



@interface BFEditTodInfoController ()

@property(nonatomic,strong) UITextView *textView;

@property(nonatomic,strong) UILabel *noticeL;

@end

@implementation BFEditTodInfoController

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.navView.navTitleL.text = @"编辑群公告";
    
    self.textView.text = self.roomInfo.intro;
    [self.view addSubview:self.textView];
    
    self.textView.editable = self.roomInfo.action_tod;
    if (self.roomInfo.action_tod) {
        [self.textView becomeFirstResponder];
        [self.navView.rightButton setTitle:@"提交" forState:UIControlStateNormal];
    }else {
        [self.view addSubview:self.noticeL];
    }
}

- (UITextView *)textView
{
    if (!_textView) {
        _textView = [[UITextView alloc]initWithFrame:CGRectMake(20, NavBar_Height + StatusBar_Height + 20, Screen_Width - 40, 150)];
        _textView.textColor = [UIColor d_colorWithColorLight:TText_Color dark:TText_Color_Dark];
        _textView.font = [UIFont systemFontOfSize:15];
    }
    return _textView;
}

- (UILabel *)noticeL
{
    if (!_noticeL) {
        _noticeL = [[UILabel alloc]initWithFrame:CGRectMake(0, Screen_Height - Bottom_SafeHeight - 100, Screen_Width, 20)];
        _noticeL.text =  @"----  管理员才能编辑和发布聊天室公告  ----";
        _noticeL.font = [UIFont systemFontOfSize:14];
        _noticeL.textColor = [UIColor lightGrayColor];
        _noticeL.textAlignment = NSTextAlignmentCenter;
    }
    return _noticeL;
}

- (void)nav_rightButtonClick
{
    if (self.textView.text.length == 0) return;
    [self.view endEditing:YES];
    WS(weakSelf)
    [[MSIMManager sharedInstance] editChatRoomTOD:self.textView.text toRoom_id:self.roomInfo.room_id successed:^{
        
        [MSHelper showToastSucc:@"发布公告成功"];
        weakSelf.roomInfo.intro = weakSelf.textView.text;
        if (weakSelf.editComplete) {
            weakSelf.editComplete();
        }
        [self.navigationController popViewControllerAnimated:YES];
    } failed:^(NSInteger code, NSString *desc) {
        [MSHelper showToastFail:desc];
    }];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    [self.view endEditing:YES];
}

@end

File: Swift/MSIM-Swift/ChatDetail/ChatRoom/BFEditTodInfoController.swift
//
//  BFEditTodInfoController.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/11/9.
//

import UIKit
import MSIMSDK


public class BFEditTodInfoController: BFBaseViewController {

    public var editComplete: (() -> Void)?
    
    public var roomInfo: MSGroupInfo!
    
    private lazy var textView: UITextView = {
     
        let textView = UITextView(frame: CGRect(x: 20, y: UIScreen.status_navi_height + 20, width: UIScreen.width - 40, height: 150))
        textView.textColor = UIColor.d_color(light: MSMcros.TText_Color, dark: MSMcros.TText_Color_Dark)
        textView.font = .systemFont(ofSize: 15)
        return textView
    }()
    
    private lazy var noticeL: UILabel = {
        let noticeL = UILabel(frame: CGRect(x: 0, y: UIScreen.height - UIScreen.safeAreaBottomHeight - 100, width: UIScreen.width, height: 20))
        noticeL.text = "----  管理员才能编辑和发布聊天室公告  ----"
        noticeL.font = .systemFont(ofSize: 14)
        noticeL.textColor = .lightGray
        noticeL.textAlignment = .center
        return noticeL
    }()
    
    public override func viewDidLoad() {
        super.viewDidLoad()
        
        self.title = "编辑群公告"
        
        view.addSubview(self.textView)
        self.textView.text = self.roomInfo.intro
        self.textView.isEditable = self.roomInfo.action_tod
        if self.roomInfo.action_tod {
            self.textView.becomeFirstResponder()
            navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(submit))
        }else {
            view.addSubview(self.noticeL)
        }
    }
    

    @objc private func submit() {
        if self.textView.text.count == 0 {return}
            self.view.endEditing(true)
        
        MSIMManager.sharedInstance().editChatRoomTOD(self.textView.text!, toRoom_id: self.roomInfo.room_id) {[weak self] in
            
            MSHelper.showToastSuccWithText(text: "发布公告成功")
            self?.roomInfo.intro = self?.textView.text
            self?.editComplete?()
            
            self?.navigationController?.popViewController(animated: true)
            
        } failed: { _, desc in
            MSHelper.showToastFailWithText(text: desc ?? "")
        }
    }
    
    public override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        view.endEditing(true)
    }
}
